plugin.video.thinktv/resources/lib/scraper.py

```diff
@@ -0,0 +1,277 @@
+   html = self.getRequest('http://www.pbs.org/watchlist/page/1/', None, xheaders)
+   if len(html) > 0:
+       a = json.loads(html)
+       if a.get('videos') is not None:
+          if len(a['videos']) > 0:
+             ilist = self.addMenuItem(addonLanguage(30005) ,'GM', ilist, 'pbswatchlist' , self.addonIcon, self.addonFanart, None, isFolder=True)
+   return ilist
+
+
+ def getAddonShows(self,url,ilist):
+    pgNum = ''
+    addonLanguage  = self.addon.getLocalizedString
+    self.doPBSLogin()
+    gsurl = url
+    if gsurl == 'favorites':
+        html = self.getRequest('http://www.pbs.org/favorite-shows-page/1/')
+        a = json.loads(html)
+        cats = a['content']
+    elif gsurl == 'localpbs':
+        xheaders = self.defaultHeaders.copy()
+        xheaders['X-Requested-With'] = 'XMLHttpRequest'
+        pbsol = self.addon.getSetting('pbsol')
+        html = self.getRequest('http://www.pbs.org/shows-page/0/?genre=&title=&callsign=%s&alphabetically=true' % (pbsol), None, xheaders)
+        a = json.loads(html)
+        cats = a['results']['content']
+    else:
+        genreUrl, pgNum = url.split('|',1) 
+        xheaders = self.defaultHeaders.copy()
+        xheaders['X-Requested-With'] = 'XMLHttpRequest'
+        html = self.getRequest('http://www.pbs.org/shows-page/%s/?genre=%s&title=&callsign=' % (pgNum, genreUrl), None, xheaders)
+        a = json.loads(html)
+        cats = a['results']['content']
+    for i, (b) in list(enumerate(cats, start=1)):
+        url = b['id']
+        name   = b['title']
+        thumb  = b['image']
+        if thumb == None : thumb = self.addonIcon
+        fanart = b['image']
+        if fanart == None: fanart = self.addonFanart
+        infoList = {}
+        infoList['TVShowTitle'] = name
+        infoList['Title']       = name
+        infoList['Studio'] = b.get('producer')
+        genres = b.get('genre_titles')
+        if genres != [] and genres is not None: infoList['Genre'] = genres[0]
+        infoList['Episode'] = b.get('video_count')
+        infoList['Plot'] = b.get('description')
+        if self.addon.getSetting('enable_login') == 'true':
+           if gsurl == 'favorites': contextMenu = [(addonLanguage(30006),'XBMC.Container.Update(%s?mode=DF&url=RF%s)' % (sys.argv[0], url))]
+           else:           contextMenu = [(addonLanguage(30007),'XBMC.RunPlugin(%s?mode=DF&url=AF%s)' % (sys.argv[0], url))]
+        else: contextMenu = None
+        ilist = self.addMenuItem(name,'GC', ilist, url, thumb, fanart, infoList, isFolder=True, cm=contextMenu)
+    if pgNum != '': ilist = self.addMenuItem('[COLOR blue]%s[/COLOR]' % addonLanguage(30050),'GS', ilist, genreUrl+'|'+str(int(pgNum)+1), self.addonIcon, self.addonFanart, None, isFolder=True)
+    return(ilist)
+
+
+ def getAddonCats(self,url,ilist):
+    addonLanguage  = self.addon.getLocalizedString
+    self.doPBSLogin()
+    thumb  = xbmc.getInfoLabel('ListItem.Art(thumb)')
+    fanart = xbmc.getInfoLabel('ListItem.Art(fanart)')
```

Comment:
the base64 idea is very good. I'm going to merge this for the time being and longer term work some thing generic out using base64 back in the helper routines library as it is a common problem with my add-ons. Thanks